modules/renter/contractor/contractor.go

```diff
@@ -0,0 +1,154 @@
+	LastRevisionTxn types.Transaction
+	SecretKey       crypto.SecretKey
+}
+
+// A Contractor negotiates, revises, renews, and provides access to file
+// contracts.
+type Contractor struct {
+	// dependencies
+	dialer  dialer
+	hdb     hostDB
+	log     logger
+	persist persister
+	tpool   transactionPool
+	wallet  wallet
+
+	allowance     modules.Allowance
+	blockHeight   types.BlockHeight
+	cachedAddress types.UnlockHash // to prevent excessive address creation
+	contracts     map[types.FileContractID]Contract
+	renewHeight   types.BlockHeight // height at which to renew contracts
+
+	mu sync.RWMutex
+}
+
+// Allowance returns the current allowance.
+func (c *Contractor) Allowance() modules.Allowance {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.allowance
+}
+
+// SetAllowance sets the amount of money the Contractor is allowed to spend on
+// contracts over a given time period, divided among the number of hosts
+// specified. Note that Contractor can start forming contracts as soon as
+// SetAllowance is called; that is, it may block.
+func (c *Contractor) SetAllowance(a modules.Allowance) error {
+	// sanity checks
+	if a.Hosts == 0 {
+		return errors.New("hosts must be non-zero")
+	} else if a.Period == 0 {
+		return errors.New("period must be non-zero")
+	}
+
+	// Set the allowance.
+	c.mu.Lock()
+	old := c.allowance
+	c.allowance = a
+	c.mu.Unlock()
+
+	// If this is the first time the allowance has been set, form contracts
+	// immediately.
+	if old.Hosts == 0 {
+		return c.formContracts(a)
+	}
+
+	// Otherwise, if the new allowance is "significantly different" (to be
+	// defined more precisely later), form intermediary contracts.
+	if a.Funds.Cmp(old.Funds) > 0 {
+		// not yet implemented
+		// c.formContracts(diff(a, old))
+	}
+
+	return nil
+}
+
+// Contracts returns the contracts formed by the contractor.
+func (c *Contractor) Contracts() (cs []Contract) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	for _, c := range c.contracts {
+		cs = append(cs, c)
+	}
+	return
+}
+
+// New returns a new Contractor.
+func New(cs consensusSet, wallet walletShim, tpool transactionPool, hdb hostDB, persistDir string) (*Contractor, error) {
+	// Check for nil inputs.
+	if cs == nil {
+		return nil, errNilCS
+	}
+	if wallet == nil {
+		return nil, errNilWallet
+	}
+	if tpool == nil {
+		return nil, errNilTpool
+	}
+
+	// Create the persist directory if it does not yet exist.
+	err := os.MkdirAll(persistDir, 0700)
+	if err != nil {
+		return nil, err
+	}
+	// Create the logger.
+	logger, err := newLogger(persistDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Contractor using production dependencies.
+	return newContractor(cs, &walletBridge{w: wallet}, tpool, hdb, stdDialer{}, newPersist(persistDir), logger)
+}
+
+// newContractor creates a Contractor using the provided dependencies.
+func newContractor(cs consensusSet, w wallet, tp transactionPool, hdb hostDB, d dialer, p persister, l logger) (*Contractor, error) {
```

Comment:
non-blocking issue, we're still experimenting with the best way to do dependencies